pkg/webhook/external/generic_list_options.go

```diff
@@ -0,0 +1,186 @@
+package external
+
+import (
+	"context"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/pkg/expressions"
+)
+
+func listTargetObjects(
+	ctx context.Context,
+	c client.Client,
+	project string,
+	target kargoapi.GenericWebhookTarget,
+	actionEnv map[string]any,
+) ([]client.Object, error) {
+	listOpts, err := buildListOptionsForTarget(project, target, actionEnv)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build list options: %w", err)
+	}
+	switch target.Kind {
+	case kargoapi.GenericWebhookTargetKindWarehouse:
+		warehouses := &kargoapi.WarehouseList{}
+		if err := c.List(ctx, warehouses, listOpts...); err != nil {
+			return nil, fmt.Errorf("error listing %s targets: %w", target.Kind, err)
+		}
+		if target.Name == "" {
+			return itemsToObjects(warehouses.Items), nil
+		}
+		for _, wh := range warehouses.Items {
+			if wh.Name == target.Name {
+				return []client.Object{&wh}, nil
+			}
+		}
+		return nil, nil
+	default:
+		return nil, fmt.Errorf("unsupported target kind: %q", target.Kind)
+	}
```

Comment:
I think I've mentioned this twice before, but I'd rather we didn't need to update this switch every time we add support for a different kind of resource.

Is there a reason we're still not using unstructured objects here?